Datasets

masojus opened the pull request:
See PR #96 for details. This PR is to continue that discussion and merge with changes made in `master` related to all the .NET Standard work.

Commit:
Add SnakeCaseContractResolved for deserializing responses to CLR types

Commit:
Add Datasets and relevant models

Commit:
Add a query definition model

Commit:
Add functions in KeenClient to query Datasets and their definitions

Commit:
Add unit test for QueryDatasetAsync

Commit:
Merge remote-tracking branch 'keenlabs/master' into cached-datasets

Commit:
Fix typo on private field: _cachedDatasetRelativeUrl

Commit:
Provide XML comments for all new public facing methods

Commit:
Provide XML comments for the dataset classes exposed via KeenClient

Commit:
Move IDataset into it's own file

Commit:
Clean up scope of using statements

Always using outside of namespace.

Commit:
Fix scope of SnakeCasecontractResolver

We don't need to expose this in the client.

Commit:
Rename Dataset.cs and provide serialization and validation utility

Commit:
Clean up usings

Commit:
Added class for fetching lists of dataset definitions

Commit:
Added Validation and Serialization utility for QueryDefinition

Commit:
Added Put methods to KeenHttpClient

Commit:
Implement Create, Delete and Get All for Cached Datasets

Commit:
Added sync methods for Create, Delete and List Cached Datasets

Commit:
Add XML comments to public facing methods

Commit:
Remove models which are not used

Commit:
Fix minor typo in DeleteDataset

Commit:
Add some simple tests for the new dataset methods

Commit:
Update dataset tests to test sync methods instead of async ones

Because the sync ones end up calling the async ones. This way all the dataset methods on KeenClient are covered.

Commit:
Use shared GetSerializerSettings function for getdatasetdefinition

Commit:
Add testing for the custom serialization logic

Commit:
Added teast for serializing datasetdefinitioncollection

Commit:
Merge branch 'master' into cached-datasets

Commit:
Fix bug with ListDatasetDefinition that popped up in unit tests

Commit:
Include testing for Create and Delete datasets

Also refactored the tests to be a bit neater

Commit:
Add tests for dataset error modes

Commit:
Clean up scope of imports

masojus committed:
Merge remote-tracking branch 'github/master' into jm_UpdateCachedDatasetsBranch

# Conflicts:
#	Keen/KeenConstants.cs

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+        }
+
+        /* It's important that there is only one instance of JsonSerializerSettings
+         * for each Serialization/Desierialization to keep the DatasetDefinitionConverter and QueryDefinitionConverter instances threadsafe */
```
This keeps those `_isNested` demarcated portions effectively thread-safe I think, if properly ensured, but it seems like it's easier to just let Json.net handle it. I would rather set the converters on the type/property in question where possible assuming we eliminate the complexity. That said, that may not be the most efficient in terms of performance, depending on how json.net creates/destroys those, so let's look into that.

masojus commented on Keen.NET.Test/DataSetTests_Integration.cs:
```diff
@@ -0,0 +1,170 @@
+namespace Keen.Net.Test
+{
+    [TestFixture]
+    public class DataSetTests_Integration : TestBase
```
I think we can swap `DataSetTests_Integration.cs` and `DatasetTests.cs` since, as far as our definition of "integration tests" goes, these are really unit tests and the other file has what we'd consider "unit tests." Also, minor nitpick: I want to rename to `DatasetTests_Integration.cs` with the same capitalization as the unit tests file.

masojus commented on Keen/Query/QueryDefinition.cs:
```diff
@@ -0,0 +1,118 @@
+        /// </summary>
+        public string Interval { get; set; }
+        /// <summary>
+        /// Specifies the name of a property by which to group results. Using this parameter changes the response format.
```
Properties plural

masojus commented on Keen/Query/QueryDefinition.cs:
```diff
@@ -0,0 +1,118 @@
+
+    internal static class QueryDefinitionExtensions
+    {
+        public static void ValidateForCachedDataset(this QueryDefinition query)
```
Being Dataset-specific these extensions could probably go in the `Dataset` namespace.

masojus commented on Keen/Dataset/DatasetDefinition.cs:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    /// <summary>
+    /// This is used because the PUT endpoints take a string for index_by, but return an array of strings.
```
Yeah via cURL I was able to create a Dataset with an array as the `index_by` so I suspect we can get away with just having `IndexBy` be an `IEnumerable<>`.

masojus commented on Keen/Query/QueryDefinition.cs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    /// <summary>
+    /// This is used because the PUT endpoint for a dataset take a string for group_by, but return an array of strings.
```
I've always understood that we could pass a string or list as the `group_by` parameter, so I'm not sure why `GroupBy` being an `IEnumerable<>` doesn't just work. We'll see.

masojus replied:
Yeah via cURL I had no issue PUTting a new Dataset with an array for `group_by`, e.g. `"group_by" : ["price"],` so I'm fairly certain we can have an `IEnumerable<>` here.

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+
+            var firstSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT);
+
+            if (firstSet?.Datasets == null)
```
maybe use null throw expression

masojus commented on Keen/Dataset/DatasetDefinition.cs:
```diff
@@ -0,0 +1,112 @@
+                throw new KeenException("DatasetDefinition must have a name.");
+            }
+
+            if (string.IsNullOrWhiteSpace(dataset.DisplayName))
```
We should make sure follows the rules for naming Cached Queries/Datasets: ```Names of Cached Datasets should be unique and can only contain alphanumeric characters, hypens ( - ), and underscores ( _ ) but the Display Name can be more human readable.```

masojus replied:
See Issue #130

masojus commented on Keen/Dataset/DatasetDefinition.cs:
```diff
@@ -0,0 +1,112 @@
+                throw new KeenException("DatasetDefinition must contain a query to be cached.");
+            }
+
+            dataset.Query.ValidateForCachedDataset();
```
Both `DatasetDefinition` and `QueryDefinition` need more validation--e.g. `timeframe` must be relative, the `timeframe` units must match the `interval`, `index_by` cannot match a `group_by` property, etc.

masojus replied:
See Issue #130

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+            return response;
+        }
+
+        public async Task<DatasetDefinition> Definition(string datasetName)
```
Actually `GetDefinitionAsync()` and the other methods should change to be `GetResultsAsync()`, etc.

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+
+            do
+            {
+                var nextSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT, allDefinitions.Last().DatasetName);
```
This doesn't seem right. It looks like in the best case scenario, if there are <100 datasets cached, we'll at least dispatch a GET to retrieve no datasets for no reason, and in worst case, we'll throw in the next few lines because we'll get back nothing. I may be missing something, but it seems this can only work if there are >= 101 dataset definitions to retrieve. We should instead be careful to only request another page if it's actually necessary.

masojus replied:
However, the `ListAllDefinitions_Success()` test uses `ListDatasetDefinitions.json` which doesn't have 100 datasets, so I'll have to take a closer look.

masojus replied:
I see, I misread the `if (firstSet.Count <= ...)` check, which should handle this situation.

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+
+                allDefinitions.AddRange(nextSet.Datasets);
+
+            } while (firstSet.Count > allDefinitions.Count);
```
This line seems confusing without a comment, but I plan to change how this works anyway.

masojus replied:
I just wonder what happens if the first request can give us a `Count` that can change by the time we request a later page...say I request 10 defs, the `Count` is 15, so we request another 10 or 5 or whatever, but since then another couple datasets have been added. We could potentially stop if we get 20 since it's >15, even if there are now actually 30 datasets because a bunch got added.

Likewise, if we request 10, the `Count` is 15, and before we request another page, some datasets get deleted, then we request another 10, but since 5 have been deleted, we get 0 back...according to the check above we'll throw even though this is technically correct.

I'll probably open a bug to handle these two cases since they'll require some deeper testing and knowing how the back end will actually behave, plus they are somewhat corner cases, so aren't on the critical path to getting this functionality usable.

masojus replied:
See Issue #132

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (responseMessage.StatusCode == HttpStatusCode.NoContent)
```
This type of info, i.e. the knowledge that we expect a 204 when we DELETE cached queries/datasets, would also be handy to have in another layer, which is covered by Issue #55.

masojus commented on Keen/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,301 @@
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var response = JObject.Parse(responseString);
```
Here and elsewhere it might be nice to avoid parsing twice. Here we `JObject.Parse()` and then a few lines later unless there's an error condition, we `JsonConvert.DeserializeObject<>()`. I wonder if either by messing with `MissingMemberHandling` when deserializing, so that if it's an error object we still get the error info, or by using `JObject.ToObject<>()` after error checking we can do this without doing two full string parsing steps.

masojus replied:
See Issue #131

masojus commented on Keen/KeenHttpClient.cs:
```diff
@@ -164,6 +164,51 @@ public Task<HttpResponseMessage> DeleteAsync(Uri resource, string authKey)
+                throw new ArgumentNullException(nameof(content), "Unexpected empty content.");
+            }
+
+            using (var contentStream = new StreamContent(new MemoryStream(Encoding.UTF8.GetBytes(content))))
```
Most of this could probably be shared with `PostAsync()`.

masojus commented on Keen/KeenClient.cs:
```diff
@@ -970,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <param name="indexBy">The string property value you want to retrieve results by.</param>
+        /// <param name="timeframe">Limits retrieval of results to a specific portion of the Cached Dataset</param>
+        /// <returns>An instance of Newtonsoft.Json.Linq.JObject containing query results and metadata defining the cached dataset.</returns>
+        public JObject QueryDataset(string datasetName, string indexBy, string timeframe)
```
Maybe the datasets return type could at least have `Result` that might be an `IEnumerable<>` if it's grouped/interval, or some other new `QueryResult` type we could introduce across the board to not just return `string` for all the metric queries. Then it could use `[JsonExtensionData]` to pack up the remaining properties, since there's some metadata and such that might be there.

masojus replied:
This is sort of related to Issue #56, but might require a separate issue apart from that.

masojus committed:
Minor formatting/style changes.

masojus committed:
No need for a `SnakeCaseContractResolver` since Json.NET has a `SnakeCaseNamingStrategy`, so use that.

masojus committed:
More PR Feedback--no need to use these custom converters:

- The back end actually supports receiving either string or list of strings for both `index_by` and `group_by`.
- In both cases it returns a list anyway.

masojus commented on Keen/KeenClient.cs:
```diff
@@ -970,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// </summary>
+        /// <param name="limit">How many cached dataset definitions to return at a time (1-100). Defaults to 10.</param>
+        /// <param name="afterName">A cursor for use in pagination. afterName is the Cached Dataset name that defines your place in the list. 
+        /// For instance, if you make a list request and receive 100 Cached Dataset definitions, 
```
misread, but we'll tidy up all these comments later anyway. We have to do a bigger pass of comments in this file.

masojus committed:
More PR feedback:

- A few comments/whitespace/formatting issues.
- Share code between PutAsync()/PostAsync().
- Move QueryDefinition extensions that are specifically for datasets to the `Keen.Core.Dataset` namespace.

masojus committed:
Swap the contents of `DatasetTests` and `DataSetTests_Integration` and also fix capitalization.

masojus committed:
Rename methods to `*Async()` since they return `Task`s.

- Still need to add or elaborate on comments for lots of public members.

masojus committed:
Filling out some interface XML doc comments.

masojus commented:
Major remaining work is captured in Issues, and we should port this all to NetStandard before release.

masojus committed:
Port Datasets to NetStandard:

- At this point all the tests pass, at least.

masojus commented:
Interesting, some of the new tests failed in Travis CI. They failed in the same way on Linux and MacOS, and it seems to be a `UriFormatException` so it could be a form of path separator issue between Windows and *nix OSes, or perhaps a permissions issue when trying to read the canned JSON response files, but I'll look into it.

masojus committed:
This should make tests be able to load those JSON files from a local path in Windows, Linux and MacOS without issue.

Shalelol commented:
Hey @masojus, thanks for all the feedback, I see you've started working on majority of the problems you pointed out. 

Please let me know if you need anything from me. We're currently in the middle of the final stages of a feature release this week, but I will be able to help out if there is anything outstanding after this week.

masojus commented:
Hi @Shalelol. Yeah I think it's close to ready. If you have any PR feedback on this latest revision, I'd love to hear it. Otherwise I'll probably merge this in the next few days.

Shalelol commented:
Hey @masojus,

Just had a quick review of your changes and it's looking really good 🔥. I'm not sure how I missed the `SnakeCaseContractResolver` already available in Json.NET. I would have picked it up if I had just read the comments on the [snippet I used](https://gist.github.com/crallen/9238178). 

Also, did the datasets API always accept `group_by` as a string? I remember having issues sending them as strings, though it may have been something else. Looking back I wish I hadn't gone through the effort of creating those custom converters and jumping through hoops for nothing 😂.

Just one small thing I noticed, though this might be better as a new issue:

- Consider consolidating versioning across projects for `Newtonsoft.Json`, `NUnit` and `Moq`. Json.NET for example has both versions `9.0.1` and `10.0.3` in the solution.

masojus commented:
Thanks @Shalelol!

I'll double check `index_by` and `group_by`, but I know the API Reference isn't crystal clear about what's actually accepted (something I'll get fixed as soon as I can), so I had to try it out. I'm not sure if/when it changed.

We went about trying to consolidate versions of Json.NET awhile back (https://github.com/keenlabs/keen-sdk-net/pull/44), but since then we've started transitioning to .NET Standard 2.0, and bumping versions of some of the dependencies only in some places, knowing we plan to eliminate the 3.5- and 4.5-specific projects. Once those are gone I plan to have all those dependencies at the same version 👍

baumatron commented on Keen.NetStandard/Dataset/QueryDefinitionExtensions.cs:
```diff
@@ -0,0 +1,31 @@
+                throw new KeenException("QueryDefinition must specify a timeframe");
+            }
+
+            if (string.IsNullOrWhiteSpace(query.Interval))
```
Is interval actually required?

Shalelol replied:
It's required by the API when creating a cached dataset. Perhaps the name of the method (`ValidateForCachedDataset`) is not clear enough.

masojus replied:
Yeah the docs I think aren't clear on this, which is another area for improvement.

masojus replied:
The docs say this:
`Currently Cached Dataset query definitions require Relative Timeframes and the unit must match the interval, `
...and this:
`Currently supports Relative Timeframes. Interval should match; eg. this_90_day and daily`
...which implies the interval should be there, but could be clearer.

baumatron commented on Keen.NetStandard/KeenClient.cs:
```diff
@@ -972,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <param name="indexBy">The string property value you want to retrieve results by.</param>
+        /// <param name="timeframe">Limits retrieval of results to a specific portion of the Cached Dataset</param>
+        /// <returns>An instance of Newtonsoft.Json.Linq.JObject containing query results and metadata defining the cached dataset.</returns>
+        public JObject QueryDataset(string datasetName, string indexBy, string timeframe)
```
Is there really a need to provide this method? If clients really want to block they can call `.Result` on the async method. Same comment for all the other non-async methods here that call `.Result` on an async method.

Shalelol replied:
Using `.Result` in Asp.Net requests that require the http context can deadlock the thread, this is why most C# APIs offer both a sync and async alternative. While the implementation for both the Sync and Async methods here are identical, they are safe because of the use of `.ConfigureAwait(false)`. The existence of the sync method implicitly informs the person consuming `KeenClient` that the method is safe. Using `.Result` on a foreign code base from within an ASP.Net application is considered a risk.

That being said, QueryDatasetAsync is missing `ConfigureAwait(false)`, I'll add the appropriate comments there and anywhere else I can find.

masojus replied:
Yeah anything that handles the sync context the way ASP does is subject to this if you haven't configured the awaiter, so this is at least self-documenting in that sense. I know S. Toub would have some words  on the subject, but I think this should stay. The real issue I have here is that I think debugging is a little tough with the exceptions we get in these wrappers, which I've been meaning to look into.

baumatron replied:
Makes sense!

baumatron commented on Keen.NetStandard/KeenClient.cs:
```diff
@@ -972,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <returns>An instance of Keen.Core.Dataset.DatasetDefinitionCollection containing the total count, next page url and list of DatasetDefinitions.</returns>
+        public Task<DatasetDefinitionCollection> ListDatasetDefinitionsAsync(int limit = 10, string afterName = null)
+        {
+            return Datasets.ListDefinitionsAsync(limit, afterName);
```
Another reason not to have the Datasets object is the extra glue code needed to call into it. The implementation could be here or in a partial class.

baumatron commented on Keen.NetStandard/Query/QueryDefinition.cs:
```diff
@@ -0,0 +1,54 @@
+    /// <summary>
+    /// Holds information describing the query that is cached within a cached dataset.
+    /// </summary>
+    public class QueryDefinition
```
I'd like to see this eventually become how we generally define a query and pass parameters to and around within the SDK.

masojus replied:
Agree

masojus replied:
Especially as a part of fixing Issues #32 and #42 my plan has been to do this since once we move to POSTing all queries it'll be cleaner.

baumatron commented on Keen.NetStandard.Test/DatasetTests_Integration.cs:
```diff
@@ -0,0 +1,434 @@
+            var dataset = client.QueryDataset(_datasetName, _indexBy, _timeframe);
+
+            Assert.IsNotNull(dataset);
+            Assert.IsNotNull(dataset["result"]);
```
Would be great to check the values actually returned for the query.

baumatron commented on Keen.NetStandard.Test/DatasetTests_Integration.cs:
```diff
@@ -0,0 +1,434 @@
+            var apiResponse = File.ReadAllText($"{GetApiResponsesPath()}/GetDatasetResults.json");
+            IKeenHttpClientProvider httpClientProvider = null;
+
+            if (UseMocks)
```
I'd like to see `UseMocks` go away.

baumatron commented on Keen.NetStandard.Test/DatasetTests_Integration.cs:
```diff
@@ -0,0 +1,434 @@
+            Assert.IsNotNull(datasetCollection);
+            Assert.IsNotNull(datasetCollection.Datasets);
+            Assert.IsTrue(datasetCollection.Datasets.Any());
+            Assert.IsTrue(!string.IsNullOrWhiteSpace(datasetCollection.NextPageUrl));
```
Again, would be great to verification of values instead of just checking to see that anything is there.

masojus replied:
Added a blurb to Issue #130 about this.

baumatron commented on Keen.NetStandard.Test/DatasetTests_Integration.cs:
```diff
@@ -0,0 +1,434 @@
+        {
+            var dataset = new DatasetDefinition();
+
+            Assert.Throws<KeenException>(() => dataset.Validate());
```
This test seems like it doesn't quite cover all validation. For example it doesn't really test that AnalysisType being unset will ever throw. Maybe separate tests for each invalid condition would be warranted? Maybe start with a valid dataset and then modify only one field at a time per-test and assert that it throws?

masojus replied:
Added a blurb to Issue #130 about this.

baumatron reviewed (changes requested):
Overall looks really good in the context of the existing SDK design! A lot of my comments are about could be done in the future for a better overall SDK design. A few smaller things should probably be addressed though like encoding url parameters and a bit of duplicated code.

Shalelol commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+            }
+
+            // This throws if dataset is not valid.
+            dataset.Validate();
```
Yeah all objects used with JSON.net require a public empty constructor.

Shalelol commented:
@baumatron I agree with your comments regarding the design of this project. It was quite strange to for me to come into and start working, and was quite different from most of the other C# projects I have worked on, but I tried to keep my code consistent with what was already there. 

Some really simple changes that could be made to drastically increase the ease of contributing to this project are:
- Moving all classes that only store data and have no logic into a separate folder (or project) called "Models" or something. At the moment they're mixed in with the code and it's very hard to find where all the actual logic is.
- Offloading all the work around building the URL and Query Parameters as well as checking the HttpStatus code to the `KeenHttpClient` and then extending the `KeenHttpClient` into a `QueriesHttpClient`, `DatasetHttpClient`, `EventCollectionHttpClient`, etc, so things like building urls for the API endpoints are not tangled up with other business logic.

Doing this would result in having a series of classes sitting next to each other (Queries, Datasets, Event, EventCache, EventCollection) whose roles are very clear, and whose code is very straight to the point, with all non-business logic nonsense abstracted away (preferably in some generic way, as you said) via the HttpClients, and potentially some helpers.

masojus commented:
@Shalelol Agreed. @baumatron and I also inherited the bulk of that code, and we created Issues #55 and #56 to basically cover exactly those points, so we understand your sentiment and are glad to hear our thoughts validated 😄

Shalelol commented on Keen.NetStandard/KeenClient.cs:
```diff
@@ -972,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <returns>An instance of Newtonsoft.Json.Linq.JObject containing query results and metadata defining the cached dataset.</returns>
+        public async Task<JObject> QueryDatasetAsync(string datasetName, string indexBy, string timeframe)
+        {
+            return await Datasets.GetResultsAsync(datasetName, indexBy, timeframe);
```
This needs to call `.ConfigureAwait(false)`

masojus replied:
Will do here and elsewhere. I fixed some down in the Datasets class but missed these 😞

Shalelol commented on Keen.NetStandard/KeenClient.cs:
```diff
@@ -972,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <returns>An instance of Keen.Core.Dataset.DatasetDefinition containing metadata about your cached dataset.</returns>
+        public async Task<DatasetDefinition> GetDatasetDefinitionAsync(string datasetName)
+        {
+            return await Datasets.GetDefinitionAsync(datasetName);
```
This needs to call `.ConfigureAwait(false)`

masojus replied:
Here and elsewhere I just return the `Task` directly since there's no need to `await` and no need to configure an awaiter--in any case internally our impls configure the awaiter, and if 3rd party impls don't, there's not much we can do about it.

Shalelol commented on Keen/KeenClient.cs:
```diff
@@ -970,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <returns>An instance of Newtonsoft.Json.Linq.JObject containing query results and metadata defining the cached dataset.</returns>
+        public async Task<JObject> QueryDatasetAsync(string datasetName, string indexBy, string timeframe)
+        {
+            return await Datasets.GetResultsAsync(datasetName, indexBy, timeframe);
```
This needs to call `.ConfigureAwait(false)`

Shalelol commented on Keen/KeenClient.cs:
```diff
@@ -970,5 +973,177 @@ public JObject Query(string queryName, Dictionary<string, string> parms)
+        /// <returns>An instance of Keen.Core.Dataset.DatasetDefinition containing metadata about your cached dataset.</returns>
+        public async Task<DatasetDefinition> GetDatasetDefinitionAsync(string datasetName)
+        {
+            return await Datasets.GetDefinitionAsync(datasetName);
```
This needs to call `.ConfigureAwait(false)`

Shalelol reviewed (commented):
Just added some comments to places where I forgot to add `.ConfigureAwait(false)` in order to keep the async methods context safe.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+    /// </summary>
+    internal class Datasets : IDataset
+    {
+        private const int MAX_DATASET_DEFINITION_LIST_LIMIT = 100;
```
.NET conventions I think tend to suggest PascalCasing for *public* constants but generally (AFAIK) don't really lay down the law w.r.t. privates. I'm not sure if it's my Java or C++, but I do this too. I'll switch to PascalCasing though, which is also inline with the CoreCLR's editorconfig, so I need to go make sure that matches before I merge the editorconfig.

masojus replied:
I've fixed this locally along with editorconfig changes.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var responseMsg = await _keenHttpClient
```
Issue #55 and #56

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var response = JObject.Parse(responseString);
```
Issue #55 and #56

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+
+            do
+            {
+                var nextSet = await ListDefinitionsAsync(MAX_DATASET_DEFINITION_LIST_LIMIT,
```
I created Issue #132 to cover a few corner cases here.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (HttpStatusCode.NoContent == responseMsg.StatusCode)
+            {
```
Personal preference, but I'm in agreement.

baumatron replied:
Agree it is style and not critical.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+
+        public async Task<DatasetDefinition> CreateDatasetAsync(DatasetDefinition dataset)
+        {
+            if (string.IsNullOrWhiteSpace(_masterKey))
```
Part of it is because yeah, unfortunately it's created even if not used (ugh...), a lot of what's in Issues #55 and #56 comes out of PR #51, but also, getting a Dataset definitions or results only requires a read key.

masojus replied:
I'd considered creating an `Attribute` to describe key-ness for these APIs the way we use decorators in the Python SDK, but haven't gotten around to looking at how well that'd work.

masojus replied:
I noticed that while ReadKey is accepted by the back end for getting results/defs, our `Datasets` impl was demanding MasterKey in those cases, so I fixed that locally, and it'll be part of the final version of this PR.

baumatron replied:
I'm curious how that might work. Can `Attribute`s work like function decorators/closures, or does some calling code have to check for the attribute and then do something?

masojus replied:
They aren't, I think, as powerful as a Python decorator, but more like Java annotations, so there'd be runtime-emitted info in the metadata that would allow some wrapper code to reflect on the method in question and decide if the proper keys are set, and how/whether to proceed. We can't fail compile like we could with a full-on metaprogramming solution in C++, but it'd be more declarative and allow us to share that key-checking logic across classes that have no common ancestor in the inheritance hierarchy.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+            }
+
+            if (string.IsNullOrWhiteSpace(prjSettings.KeenUrl) ||
+                !Uri.IsWellFormedUriString(prjSettings.KeenUrl, UriKind.Absolute))
```
Yeah I'm actually not sure why I didn't do that during PR #51 or whenever it was that I added these lines of code to `Queries`/et. al.

masojus replied:
I was playing with this and realized that part of the problem is you can implement `IProjectSettings` in 3rd party code. That said, in theory since our impls of `Datasets`/`Queries`/`Event`/etc. have `internal` ctors, you can't directly construct them, so we *could* just call it sufficient to validate the project settings in the `KeenClient` ctor. However, I think this is better off being fixed as part of Issue #56 since we should consolidate a bunch of the code in all those impls.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+    /// Datasets implements the IDataset interface which represents the Keen.IO Cached Datasets
+    /// API methods.
+    /// </summary>
+    internal class Datasets : IDataset
```
Issues #55 and #56

baumatron replied:
Yep, not a comment on this change really... It's preferable that this change follow the existing design of the SDK.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+
+            var datasetResultsUrl = $"{GetDatasetUrl(datasetName)}/results";
+
+            var url = $"{datasetResultsUrl}?index_by={indexBy}&timeframe={timeframe}";
```
Yeah here and elsewhere we need to verify this before merging.

masojus replied:
`index_by` can be any JSOn member name, I think, which in theory can be all sorts of things that need encoding.
`timeframe` when retrieving results can be an absolute timeframe (not just relative) I think, so it could have things like a colon, so also needs encoding.
`limit` shouldn't need encoding since it's just an int.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+            // This throws if dataset is not valid.
+            dataset.Validate();
+
+            var content = JsonConvert.SerializeObject(dataset, SERIALIZER_SETTINGS);
```
Public constants I think tend to be PascalCasing. That's just C++ and Java poisoning my brain I bet.

masojus replied:
Fill fix in another PR with editorconfig changes.

masojus commented on Keen.NET.Test/DataSetTests_Integration.cs:
```diff
@@ -0,0 +1,435 @@
+            var apiResponse = File.ReadAllText($"{GetApiResponsesPath()}/GetDatasetResults.json");
+            IKeenHttpClientProvider httpClientProvider = null;
+
+            if (UseMocks)
```
Yeah I thought maybe we'd look at this more closely with Issue #37, but I'd say we close that issue as fixed with PR #127 and instead open a more specific issue detailing what to do about `UseMocks` in the future.

baumatron replied:
Seems fine.

masojus replied:
Issue #133

baumatron commented:
@Shalelol All great ideas! I also have been trying to keep with the existing design of the SDK until there's an opportunity to do a larger refactor as you describe. Would love to see that in the future.

masojus committed:
Merge branch 'master' into jm_UpdateCachedDatasetsBranch

masojus committed:
Responding to some PR Feedback:

- Regarding *Async() methods in KeenClient.cs, for simple forwarding methods, just return the task directly. No need to await or configure the awaiter...either way, internally our impls configure awaiters properly everywhere, and a 3rd party impl of, say IQueries could fail to do so and cause blocking on our task to potentially deadlock anyway, depending on the sync ctx. Anywhere there are calls to multiple async APIs we go ahead and await/configure as expected.
- One actual ramification of returning tasks is that in a few places where we validate params, we'll throw synchronously now instead of when the task is awaited, so potentially if you call us and store the task to await later, the client code's try/catch could be in the wrong place, but that seems like it's not that huge a deal as long as we bump the minor version (still pre-v1.0).
- Some of the Datasets code was incorrectly demanding a MasterKey when it really only needs a ReadKey.
- Changing to ReadKey showed me that some of the tests that expect throwing situations wouldn't catch the wrong key, because we throw the same exception for the server returning 500 as for an incorrect key (KeenException). This could mean we should throw different exceptions, which would make testing this easier (and potentially make coding against the SDK easier). It could also be fixed in tests by checking for certain words in the error message, but I figured it was more straightforward and clear to separate that particular throwing condition into its own test for each of the create/result/definition(s)/delete functions.
- Changed GetQueries()->GetQueriesAsync(), since it's public and should indicate that it returns an awaitable.

masojus committed:
In response to PR feedback, escape some of the URL query string params for Datasets.

masojus commented on Keen.NetStandard/Dataset/Datasets.cs:
```diff
@@ -0,0 +1,306 @@
+
+            if (!string.IsNullOrWhiteSpace(afterName))
+            {
+                datasetResultsUrl += $"&after_name={afterName}";
```
`after_name` should be a valid Dataset name, which can only be alphanumerics, '_' and '-', so we won't need to escape here. The back end will only accept valid names, and as part of fixing Issue #130 we should be validating these coming in before we get here.
